File: mobile/android/focus-android/app/src/main/java/org/mozilla/focus/session/ui/SessionsAdapter.java
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.mozilla.focus.session.ui;

import android.arch.lifecycle.Observer;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.TextView;

import org.mozilla.focus.session.Session;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Adapter implementation to show a list of active browsing sessions and an "erase" button at the end.
 */
public class SessionsAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> implements Observer<List<Session>> {
    private final SessionsSheetFragment fragment;
    private List<Session> sessions;

    /* package */ SessionsAdapter(SessionsSheetFragment fragment) {
        this.fragment = fragment;
        this.sessions = Collections.emptyList();
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        final LayoutInflater inflater = LayoutInflater.from(parent.getContext());

        switch (viewType) {
            case EraseViewHolder.LAYOUT_ID:
                return new EraseViewHolder(
                        fragment,
                        inflater.inflate(EraseViewHolder.LAYOUT_ID, parent, false));
            case SessionViewHolder.LAYOUT_ID:
                return new SessionViewHolder(
                        fragment,
                        (TextView) inflater.inflate(SessionViewHolder.LAYOUT_ID, parent, false));
            default:
                throw new IllegalStateException("Unknown viewType");
        }
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        switch (holder.getItemViewType()) {
            case EraseViewHolder.LAYOUT_ID:
                // Nothing to do here.
                break;
            case SessionViewHolder.LAYOUT_ID:
                ((SessionViewHolder) holder).bind(sessions.get(position));
                break;
            default:
                throw new IllegalStateException("Unknown viewType");
        }
    }

    @Override
    public int getItemViewType(int position) {
        if (isErasePosition(position)) {
            return EraseViewHolder.LAYOUT_ID;
        } else {
            return SessionViewHolder.LAYOUT_ID;
        }
    }

    private boolean isErasePosition(int position) {
        return position == sessions.size();
    }

    @Override
    public int getItemCount() {
        return sessions.size() + 1;
    }

    @Override
    public void onChanged(List<Session> sessions) {
        this.sessions = new ArrayList<>(sessions);
        notifyDataSetChanged();
    }
}

File: mobile/android/focus-android/app/src/main/java/org/mozilla/focus/session/ui/SessionsAdapter.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.mozilla.focus.session.ui

import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.TextView
import mozilla.components.browser.session.Session
import mozilla.components.browser.session.SessionManager
import org.mozilla.focus.ext.requireComponents

/**
 * Adapter implementation to show a list of active browsing sessions and an "erase" button at the end.
 */
class SessionsAdapter internal constructor(
    private val fragment: SessionsSheetFragment,
    private var sessions: List<Session> = emptyList()
) : RecyclerView.Adapter<RecyclerView.ViewHolder>(), SessionManager.Observer {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val inflater = LayoutInflater.from(parent.context)

        return when (viewType) {
            EraseViewHolder.LAYOUT_ID -> EraseViewHolder(
                fragment,
                inflater.inflate(EraseViewHolder.LAYOUT_ID, parent, false)
            )
            SessionViewHolder.LAYOUT_ID -> SessionViewHolder(
                fragment,
                inflater.inflate(SessionViewHolder.LAYOUT_ID, parent, false) as TextView
            )
            else -> throw IllegalStateException("Unknown viewType")
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (holder.itemViewType) {
            EraseViewHolder.LAYOUT_ID -> { /* Nothing to do */ }
            SessionViewHolder.LAYOUT_ID -> (holder as SessionViewHolder).bind(sessions[position])
            else -> throw IllegalStateException("Unknown viewType")
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (isErasePosition(position)) {
            EraseViewHolder.LAYOUT_ID
        } else {
            SessionViewHolder.LAYOUT_ID
        }
    }

    private fun isErasePosition(position: Int): Boolean {
        return position == sessions.size
    }

    override fun getItemCount(): Int {
        return sessions.size + 1
    }

    override fun onSessionAdded(session: Session) {
        onUpdate(fragment.requireComponents.sessionManager.sessions)
    }

    override fun onSessionRemoved(session: Session) {
        onUpdate(fragment.requireComponents.sessionManager.sessions)
    }

    override fun onSessionSelected(session: Session) {
        onUpdate(fragment.requireComponents.sessionManager.sessions)
    }

    override fun onAllSessionsRemoved() {
        onUpdate(fragment.requireComponents.sessionManager.sessions)
    }

    private fun onUpdate(sessions: List<Session>) {
        this.sessions = sessions
        notifyDataSetChanged()
    }
}
